File: geocoding_darwin/darwin/Classes/GeocodingPlugin.m
//
//  GeocodingPlugin.m
//  geocoding
//
//  Created by Maurits van Beusekom on 07/06/2020.
//

#import "CLPlacemarkExtensions.h"
#import "GeocodingHandler.h"
#import "GeocodingPlugin.h"

@implementation GeocodingPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"flutter.baseflow.com/geocoding"
                                     binaryMessenger:[registrar messenger]];
    GeocodingPlugin* instance = [[GeocodingPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"locationFromAddress" isEqualToString:call.method]) {
        NSString* address = call.arguments[@"address"];
        
        GeocodingHandler* handler = [[GeocodingHandler alloc] init];
        [handler geocodeFromAddress:address
                             locale:[GeocodingPlugin parseLocale: call.arguments]
                            success:^(NSArray<CLPlacemark *> * placemarks) {
            result([GeocodingPlugin toLocationResult: placemarks]);
        }
                            failure:^(NSString * _Nonnull errorCode, NSString * _Nonnull errorDescription) {
            result([FlutterError errorWithCode:errorCode
                                       message:errorDescription
                                       details:nil]);
        }];
    } else if ([@"placemarkFromCoordinates" isEqualToString:call.method]) {
        CLLocationDegrees latitude = ((NSNumber *) call.arguments[@"latitude"]).doubleValue;
        CLLocationDegrees longitude = ((NSNumber *) call.arguments[@"longitude"]).doubleValue;
        CLLocation* location = [[CLLocation alloc] initWithLatitude:latitude
                                                          longitude:longitude];
        
        GeocodingHandler* handler = [[GeocodingHandler alloc] init];
        [handler geocodeToAddress:location
                           locale:[GeocodingPlugin parseLocale: call.arguments]
                          success:^(NSArray<CLPlacemark *> * placemarks) {
            result([GeocodingPlugin toPlacemarkResult: placemarks]);
        }
                          failure:^(NSString * _Nonnull errorCode, NSString * _Nonnull errorDescription) {
            result([FlutterError errorWithCode:errorCode
                                       message:errorDescription
                                       details:nil]);
        }];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

+ (NSLocale*) parseLocale:(NSDictionary*)arguments {
    if (arguments[@"localeIdentifier"] == nil) {
        return nil;
    }
    
    return [[NSLocale alloc] initWithLocaleIdentifier:(NSString*) arguments[@"localeIdentifier"]];
}

+ (NSArray<NSDictionary *> *) toLocationResult:(NSArray<CLPlacemark *> *)placemarks {
    NSMutableArray<NSDictionary *> *result = [[NSMutableArray alloc] init];
    
    for (CLPlacemark *placemark in  placemarks) {
        [result addObject:[placemark toLocationDictionary]];
    }
    
    return result;
}

+ (NSArray<NSDictionary *> *) toPlacemarkResult:(NSArray<CLPlacemark *> *)placemarks {
    NSMutableArray<NSDictionary *> *result = [[NSMutableArray alloc] init];
    
    for (CLPlacemark *placemark in  placemarks) {
        [result addObject:[placemark toPlacemarkDictionary]];
    }
    
    return result;
}

@end

File: geocoding/macos/Classes/GeocodingPlugin.swift
import Cocoa
import FlutterMacOS
import CoreLocation

public class GeocodingPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter.baseflow.com/geocoding", binaryMessenger: registrar.messenger)
        let instance = GeocodingPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "locationFromAddress":
            if let args = call.arguments as? Dictionary<String, Any>, let address: String = args["address"] as? String {
                GeocodingHandler().geocodeFromAddress(
                    address: address,
                    locale: parseLocale(args),
                    success: { result(self.toLocationResult($0.placemarks)) },
                    failure: { result(FlutterError(code: $0.errorCode, message: $0.errorDescription, details: nil)) }
                )
            } else {
                result(FlutterError(code: "MISSING DATA", message: "Missing Address field", details: nil))
            }
        case "placemarkFromCoordinates":
            if let args = call.arguments as? Dictionary<String, Any>, let latitude = args["latitude"] as? NSNumber, let longitude = args["longitude"] as? NSNumber {
                let lat = CLLocationDegrees(latitude.doubleValue)
                let long = CLLocationDegrees(longitude.doubleValue)
                let location: CLLocation = CLLocation(latitude: lat, longitude: long)
                GeocodingHandler().geocodeToAddress(
                    location: location,
                    locale: parseLocale(args),
                    success: { result(self.toPlacemarkResult($0.placemarks)) },
                    failure: { result(FlutterError(code: $0.errorCode, message: $0.errorDescription, details: nil)) }
                )
            } else {
                result(FlutterError(code: "MISSING DATA", message: "Missing latitude or longitude fields", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func parseLocale(_ args: Dictionary<String, Any>) -> Locale? {
        
        guard let localeId: String = args["localeIdentifier"] as? String else {
            return nil
        }
        
        return Locale(identifier: localeId)
    }
    
    public func toLocationResult(_ placemarks: [CLPlacemark]) -> [Dictionary<String, Any>] {
        var result: [Dictionary<String, Any>] = []
        
        for placemark in placemarks {
            if let placeDict = placemark.toLocationDictionary {
                result.append(placeDict)
            }
        }
        
        return result
    }
    
    public func toPlacemarkResult(_ placemarks: [CLPlacemark]) -> [Dictionary<String, Any>] {
        var result: [Dictionary<String, Any>] = []
        
        for placemark in placemarks {
            result.append(placemark.toPlacemarkDictionary)
        }
        
        return result
    }
    
}
